components/common/src/test/java/com/hotels/styx/config/validator/ObjectValidatorTest.java

```diff
@@ -0,0 +1,655 @@
+                        + "plugins:\n"
+                        + "  all: 'x, y z'\n");
+
+        ObjectValidator validator = newDocument()
+                .subSchema("HttpPipeline", schema()
+                        .field("pipeline", string())
+                        .field("handler", string())
+                )
+                .subSchema("PluginsList", schema()
+                        .field("all", string())
+                )
+                .rootSchema(schema("")
+                        .field("httpPipeline", object("HttpPipeline"))
+                        .field("plugins", object("PluginsList")))
+                .build();
+
+        assertThat(validator.validateObject(rootObject), is(true));
+    }
+
+    @Test(expectedExceptions = InvalidSchemaException.class,
+            expectedExceptionsMessageRegExp = "No schema configured for lazy object reference 'notExists'")
+    public void documentBuilderEnsuresNamedSchemaReferences() throws Exception {
+        newDocument()
+                .subSchema("PluginsList", schema()
+                        .field("all", string())
+                )
+                .rootSchema(schema("")
+                        .field("httpPipeline", object("notExists"))
+                        .field("plugins", object("PluginsList")))
+                .build();
+    }
+
+    @Test(expectedExceptions = InvalidSchemaException.class,
+            expectedExceptionsMessageRegExp = "No schema configured for lazy object reference 'notExists'")
+    public void documentBuilderEnsuresNamedSchemaReferencesAreValidFromSubobjects() throws Exception {
+        newDocument()
+                .subSchema("PluginsList", schema()
+                        .field("all", string()))
+                .rootSchema(schema("")
+                        .field("parent", object(
+                                schema("subobject")
+                                        .field("httpPipeline", object("notExists"))
+                        ))
+                        .field("plugins", object("PluginsList")))
+                .build();
+    }
+
+    @Test(expectedExceptions = InvalidSchemaException.class,
+            expectedExceptionsMessageRegExp = "No schema configured for lazy object reference 'non-existing-HandlerConfig'")
+    public void documentBuilderEnsuresNamedSchemaReferencesAreValidFromSubSchemas() throws Exception {
+        newDocument()
+                .subSchema("HttpPipeline", schema()
+                        .field("handlers", object("non-existing-HandlerConfig"))
+
+                )
+                .subSchema("PluginsList", schema()
+                        .field("all", string())
+                )
+                .rootSchema(schema("")
+                        .field("plugins", object("PluginsList"))
+                        .field("httpPipeline", object("HttpPipeline")))
+                .build();
+    }
+
+    @Test
+    public void validatesDiscriminatedUnions() throws Exception {
```

Comment:
Please add a comment defining the term "discriminated union".

Reply:
It is already described in the Schema.java where the `DiscriminatedUnionObject` lives. But I think I'll need to move the documentation over to Schema Field class where the static builder methods are.